Core/include/Acts/Utilities/TrackHelpers.hpp

```diff
@@ -0,0 +1,339 @@
+    const Logger &logger = *getDefaultLogger("TrackSmoother", Logging::INFO)) {
+  Result<void> result = Result<void>::success();
+
+  for (const auto &track : trackContainer) {
+    auto smoothingResult =
+        smoothTrack(geoContext, trackContainer, track, logger);
+
+    if (!smoothingResult.ok()) {
+      result = smoothingResult.error();
+      continue;
+    }
+  }
+
+  return result;
+}
+
+/// @brief Find a track state for extrapolation
+///
+/// @tparam track_proxy_t The track proxy type
+///
+/// @param geoContext The geometry context
+/// @param track The track
+/// @param referenceSurface The reference surface
+/// @param strategy The extrapolation strategy
+/// @param logger The logger
+///
+/// @return The result of the search
+template <typename track_proxy_t>
+Result<std::pair<typename track_proxy_t::ConstTrackStateProxy, double>>
+findTrackStateForExtrapolation(
+    const GeometryContext &geoContext, track_proxy_t &track,
+    const Surface &referenceSurface, TrackExtrapolationStrategy strategy,
+    const Logger &logger = *getDefaultLogger("TrackExtrapolation",
+                                             Logging::INFO)) {
+  using TrackStateProxy = typename track_proxy_t::ConstTrackStateProxy;
+
+  auto findFirst = [&]() -> Result<TrackStateProxy> {
+    // TODO specialize if track is forward linked
+
+    auto result = Result<TrackStateProxy>::failure(
+        TrackExtrapolationError::CompatibleTrackStateNotFound);
+
+    for (const auto &trackState : track.trackStatesReversed()) {
+      bool isMeasurement =
+          trackState.typeFlags().test(TrackStateFlag::MeasurementFlag);
+      bool isOutlier = trackState.typeFlags().test(TrackStateFlag::OutlierFlag);
+
+      if (isMeasurement && !isOutlier) {
+        result = trackState;
+      }
+    }
+
+    return result;
+  };
+
+  auto findLast = [&]() -> Result<TrackStateProxy> {
+    for (const auto &trackState : track.trackStatesReversed()) {
+      bool isMeasurement =
+          trackState.typeFlags().test(TrackStateFlag::MeasurementFlag);
+      bool isOutlier = trackState.typeFlags().test(TrackStateFlag::OutlierFlag);
+
+      if (isMeasurement && !isOutlier) {
+        return trackState;
+      }
+    }
+
+    return Result<TrackStateProxy>::failure(
+        TrackExtrapolationError::CompatibleTrackStateNotFound);
+  };
+
+  auto intersect = [&](const TrackStateProxy &state) -> SurfaceIntersection {
+    auto freeVector = MultiTrajectoryHelpers::freeSmoothed(geoContext, state);
+
+    return referenceSurface
+        .intersect(geoContext, freeVector.template segment<3>(eFreePos0),
+                   freeVector.template segment<3>(eFreeDir0),
+                   BoundaryCheck(true), s_onSurfaceTolerance)
+        .closest();
+  };
+
+  if (strategy == TrackExtrapolationStrategy::first) {
+    ACTS_VERBOSE("looking for first track state");
+
+    auto first = findFirst();
+    if (!first.ok()) {
+      ACTS_ERROR("no first track state found");
```

Comment:
We'll have to see if these ERRORs cause problems downstream, might have to make the logging behavior configurable again.

Reply:
I think right now this can only happen if we do not have any measurements states at all. In this case we should really not try to extrapolate the track. One way to avoid this would be to use a track selector before extrapolation which could also save some CPU time